Add Ubuntu automatic installation

DaniilKl opened the pull request:
Adds Automatic installation of Ubuntu Desktop instance via `netboot.xyz` preconfigured using [`Dasharo/preseeds`](https://github.com/Dasharo/preseeds) repo.

Commit:
scripts: qemu-run.sh: increase QEMU RAM to 8G

Otherwise cloud-init will be killed by OOM-Killer during installation
via netboot.xyz with autoinstall option.

Signed-off-by: Daniil Klimuk <daniil.klimuk@3mdeb.com>

Commit:
lib: add KW for netboot.xyz

Signed-off-by: Daniil Klimuk <daniil.klimuk@3mdeb.com>

Commit:
operating-systems: ubuntu.robot: add

Adds tests for installing Ubuntu 22.04 OS.

Signed-off-by: Daniil Klimuk <daniil.klimuk@3mdeb.com>

Commit:
operating-systems: ubuntu.robot: update to Ubuntu 24.04

Signed-off-by: Daniil Klimuk <daniil.klimuk@3mdeb.com>

DaniilKl commented on operating-systems/ubuntu.robot:
```diff
@@ -0,0 +1,61 @@
+
+    # Wait for Ubuntu to be installed:
+    # 1. Increase timeout, installation cat take a while
+    Set DUT Response Timeout    120m
```
For me the longest installation time was around 20 minutes. During installation you should see a lot of installer logs on your monitor (or serial port), so in your case I guess the installation failed on chaining Ubuntu installer for iPXE.

DaniilKl commented on operating-systems/ubuntu.robot:
```diff
@@ -0,0 +1,61 @@
+    Press Enter
+
+    # Wait for Ubuntu to be installed:
+    # 1. Increase timeout, installation cat take a while
```
Amended here 1aaf36e68c683beb6705bcd5962ed08f191d4993.

DaniilKl commented on operating-systems/ubuntu.robot:
```diff
@@ -0,0 +1,61 @@
+    # 1. Increase timeout, installation cat take a while
+    Set DUT Response Timeout    120m
+    # 2. Verify that Ubuntu has been installed
+    Set Prompt For Terminal    3mdeb login:
```
Replaced here 1aaf36e68c683beb6705bcd5962ed08f191d4993, thanks.

DaniilKl commented:
Noticed some weird behavior on freshly auto-installed Ubuntu 24.04: it takes too much time to boot it (4-5 minutes). I have found following logs in `dmesg`:

```shell
(...)
[FAILED] Failed to start systemd-networkd-w… Wait for Network to be Configured.
See 'systemctl status systemd-networkd-wait-online.service' for details.
         Starting cloud-init.service - Init…t job (metadata service crawler)...
[  132.110535] cloud-init[1100]: Cloud-init v. 24.1.3-0ubuntu3.3 running 'init' at Mon, 08 Jul 2024 07:39:59 +0000. Up 132.09 seconds.
[  132.156613] cloud-init[1100]: ci-info: +++++++++++++++++++++++++++++Net device info+++++++++++++++++++++++++++++
[  132.169123] cloud-init[1100]: ci-info: +-----------+-------+-----------+-----------+-------+-------------------+
[  132.182108] cloud-init[1100]: ci-info: |   Device  |   Up  |  Address  |    Mask   | Scope |     Hw-Address    |
[  132.195117] cloud-init[1100]: ci-info: +-----------+-------+-----------+-----------+-------+-------------------+
[  132.208141] cloud-init[1100]: ci-info: |   enp3s0  | False |     .     |     .     |   .   | d8:bb:c1:8c:be:7e |
[  132.222120] cloud-init[1100]: ci-info: |     lo    |  True | 127.0.0.1 | 255.0.0.0 |  host |         .         |
[  132.236075] cloud-init[1100]: ci-info: |     lo    |  True |  ::1/128  |     .     |  host |         .         |
[  132.249070] cloud-init[1100]: ci-info: | wlp0s20f3 | False |     .     |     .     |   .   | 64:6e:e0:6b:9e:82 |
[  132.262063] cloud-init[1100]: ci-info: +-----------+-------+-----------+-----------+-------+-------------------+
[  132.275069] cloud-init[1100]: ci-info: +++++++++++++++++++Route IPv6 info+++++++++++++++++++
[  132.286024] cloud-init[1100]: ci-info: +-------+-------------+---------+-----------+-------+
[  132.297067] cloud-init[1100]: ci-info: | Route | Destination | Gateway | Interface | Flags |
[  132.309111] cloud-init[1100]: ci-info: +-------+-------------+---------+-----------+-------+
[  132.320068] cloud-init[1100]: ci-info: +-------+-------------+---------+-----------+-------+
         Starting modprobe@dm_mod.service - Load Kernel Module dm_mod...
         Starting modprobe@efi_pstore.servi… - Load Kernel Module efi_pstore...
         Starting modprobe@loop.service - Load Kernel Module loop...
         Starting systemd-networkd-wait-onl…ait for Network to be Configured...
(...)
```

It seems it tries to start `systemd-networkd0wait-online` for 4 minutes and fails, then it continues to boot system. What is interesting, though, the system boots with network connection configured.

I do not know what is going on but two facts are certain: the system boots significantly slower and Ubuntu 22.04 did not have such problems.

pietrushnic committed:
add lib/bios/seabios.robot, remove resource menus.robot

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
platform-configs/include/default.robot: disable OPTIONS_LIB, which depends on menus.robot and conflict seabios

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
platform-configs/qemu-selftests.robot: change configuration to one used by seabios

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
scripts/ci/qemu-run.sh: give ability to provide QEMU_FW_FILE through env var

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
scripts/ci/qemu-run.sh: enabled kvm only if it supported

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
scripts/ci/qemu-self-test-seabios.sh: initial commit

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
self-tests/setup-and-boot-seabios.robot: initial commit

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
scripts: add support for Dasharo (coreboot+SeaBIOS) for QEMU Q35

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/seabios.robot: add more keywords and its tests

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
add support for BIOS_LIB uefi or seabios

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
.github/workflows/qemu-self-test-seabios.yml: initial commit

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

macpijan committed:
platform-configs/qemu: swap release and selftests configs

It appers to me that the selftest config should have more menus
enabled, than the "release" config, supporting only "reasonable"
menus.

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
platform-configs/qemu-selftests-uefi.robot: use common part from qemu,robot

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

pietrushnic committed:
keywords.robot: do not import lib/bios/menus.robot twice

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
platform-configs/qemu-selftests-uefi.robot: fix resource order

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
platform-configs,scripts: make BIOS_LIB=uefi default and override only in qemu seabios config

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
platform-configs/qemu-selftests-seabios.robot: import common qemu.robot resource

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
docs,scripts: keep firmware option for qemu-run.sh as backward compatibility

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
rename lib/bios/menus.robot -> lib/bios/edk2.robot

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
use platform-defined FW_STRING instead of {SEABIOS,TIANOCORE}_STRING

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/common.robot: initial support for common Enter Boot Menu kwd

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
add unified Enter Boot Menu and Return Construction

This commit replace:
- Enter Boot Menu Tianocore And Return Construction
- Enter Boot Menu SeaBIOS And Return Construction
with one keyword Enter Boot Menu and Return Construction, which is
provided by lib/bios/common.robot

It also move Parse Menu Snapshot Into Construction to common lib.

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/seabios.robot: remove unused Get IPXE Boot Menu Construction

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/seabios.robot: remove unused Get RTC Clock Submenu Construction

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
unify Enter Setup Menu kwd for uefi and seabios

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios: user Get Menu Construction instead of SeaBIOS dedicated wrapper

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/seabios.robot: remove unused keywords

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/edk2.robot: remove Get Menu Consutruction use common implmentation

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/seabios.robot: align API with edk2 lib

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios: move Get Index Of Matching Option In Menu to common

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
seabios: switch from using Enter Menu From Snapshot to Enter Submenu ... kwd

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios: move some keywords to common lib

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
start enable network-boot tests in sebios

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
Add functionality for running tests on a remote machine

This commit introduces a new feature that allows users to run tests on a
remote machine. This is particularly useful when running QEMU on the
local machine is not desirable due to security concerns or memory
limitations.

The changes include updates to the documentation to explain the new
feature and a new script, `remote-runner.sh`, which handles the
synchronization of a local directory to a remote machine and the
execution of QEMU based on parameters on the remote machine.

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

filipleple committed:
lib/bios: move Boot System Or From Connected Disk to common lib

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
platform-configs/optiplex-7010-seabios.robot: add

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
lib/bios: make Boot System Or From Connected Disk more generic

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

pietrushnic commented:
@macpijan, why is it stuck? I need HDD to test Dasharo (coreboot+SeaBIOS), and AFAIK, I have no way of installing OS in QEMU other than manually. This is weird since it essentially limits validation.

pietrushnic committed:
scripts: refactor qemu-run.sh and remote-runner.sh scripts

This commit refactors the qemu-run.sh and remote-runner.sh scripts to
improve their flexibility and usability.

In qemu-run.sh, the ACTION parameter has been split into ACTION and
FIRMWARE parameters. This allows the user to specify the action (os or
os_install) and the firmware (uefi or seabios) separately, providing
more flexibility in the configuration of the QEMU instance. The usage
instructions have been updated accordingly. Now we are not limited to
run tests or install only on uefi.

In remote-runner.sh, the number of required parameters has been
increased from 4 to 5 to accommodate the new FIRMWARE parameter in
qemu-run.sh. The usage instructions and error messages have been updated
to reflect this change.

Additionally, the shebang in remote-runner.sh has been changed from
different systems.

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
scripts/ci/qemu-run.sh: update to ubuntu 24.04.1

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/edk2.robot: fix rebase issues

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
.github/workflows: add support for new qemu-run.sh interface

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/edk2.robot: further cleanup after cleanup

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
lib/bios/common.robot: improve matching pattern

It is very interesting that incorrect pattern fails because of
underlying logic:

KEYWORD ${matches} = BuiltIn . Run Keyword And Return Status Should Match, ${item}, *${option}*
Documentation:

23:12:21.311 	TRACE 	Arguments: [ '2. AHCI/0: QEMU HARDDISK ATA-7 Hard-Disk (20480 MiBytes)' | '*[setup]*' ]
23:12:21.311 	TRACE 	Return: None
23:12:21.311 	TRACE 	Return: True
23:12:21.311 	INFO 	${matches} = True

So this one return True, but following one return False:

KEYWORD ${matches} = BuiltIn . Run Keyword And Return Status Should Match, ${item}, *${option}*
Documentation:

Fails if the given string does not match the given pattern.
Start / End / Elapsed: 	20241216 23:12:21.307 / 20241216 23:12:21.309 / 00:00:00.002
23:12:21.307 	TRACE 	Arguments: [ '1. DVD/CD [AHCI/2: QEMU DVD-ROM ATAPI-4 DVD/CD]' | '*[setup]*' ]
23:12:21.308 	FAIL 	'1. DVD/CD [AHCI/2: QEMU DVD-ROM ATAPI-4 DVD/CD]' does not match '*[setup]*'
23:12:21.309 	DEBUG 	Traceback (most recent call last):
  None
AssertionError: '1. DVD/CD [AHCI/2: QEMU DVD-ROM ATAPI-4 DVD/CD]' does not match '*[setup]*'
23:12:21.309 	TRACE 	Return: False
23:12:21.309 	INFO 	${matches} = False

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
scripts/ci/qemu-run.sh: move use of virtio-net to uefi params

This patch fix the problem where virtio-net by delivering its pxe rom
interfere with the fact that seabios needs empty romfile and use built
in iPXE to do network boot and work with developed tests.

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
scripts/ci/remote-runner.sh: silence pip and rsync

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
scripts/ci/remote-runner.sh: make sure qemu-run gets the same parameters in env variables no mater the QEMU_FW_FILE

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
dasharo-compatibility/network-boot.robot: correct PXE003.01 goals

The current PXE003.01 test, which is supposed to check the functionality
of autoboot, does not perform as expected. The test passes based on the
appearance of a follow-up menu "Please select boot device", which varies
under SeaBIOS.

The issue arises from the lack of DHCP returning the necessary
information, leading to the "Nothing to boot iPXE error:
https://ipxe.org/err/2d03e1".

To correct this, DHCP should return a root-path and possibly combine
it with TFTP to retrieve the file. This approach is feasible in a QEMU
environment with customizable DHCP and TFTP. However, it might be more
challenging in a lab setting.

To fix above issue this commit change description of the test and make
it work for both uefi and seabios firmware.

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

macpijan commented:
Because no one has successfully repeated this yet. And for installing on hardware we mostly use this method which has worked so far: https://github.com/Dasharo/preseeds?tab=readme-ov-file#ubuntu

DaniilKl commented:
> Because no one has successfully repeated this yet.

I have used it several times while running tests on several platforms. If smbd met any problem here - then the problem should be reported and handled accordingly.

macpijan commented:
I think I tried once or twice and reported once. Also asked @dawidlesiak many times to do so.

I guess the limitation is here that in 90% of cases we need Ubuntu to be installed alongside of Windows, not on the whole disk?

pietrushnic commented:
> I guess the limitation is here that in 90% of cases we need Ubuntu to be installed alongside of Windows, not on the whole disk?

Why? Where do we get that limitation from?

I want to rebase this code on my seabios changes and try it in QEMU; what would be the best approach to that?

DaniilKl commented:
> I want to rebase this code on my seabios changes and try it in QEMU; what would be the best approach to that?

Just rebase and run it using standard OSFV workflow for Qemu.

Hope there were no changes in netboot.xyz menu since last time I used this workflow. Otherwise, you may have problems, because, as you can see from code, currently netboot.xyz menu is hardcoded.

pietrushnic committed:
scripts/ci/qemu-self-test-seabios.sh: enable OS tests on SeaBIOS

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

macpijan commented:
> Why? Where do we get that limitation from?

We need W11 for testing on nearly all (except for PC Engines) platforms. There is usually one disk. W11 autoinstaller needs to go first as it erases the whole disk. Then the ubuntu semi-auto installer needs to install Ubuntu alongside of Windows with manual partition selection, which we failed to automate after several attempts IIRC: https://github.com/Dasharo/preseeds/issues/5

This MR also does not solve it.

macpijan commented:
If testing this on APU, be warned that at least on some of them it would not work, as the Ubuntu network ISO is already almost 3GB and would fail to load over netboot.xyz on 2GB units at least.

pietrushnic committed:
lib/netbootxyz-lib.robot: add support for common bios lib

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
operating-systems/ubuntu.robot: sonoff and rtecrtl api are no longer used

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
scripts/ci/qemu-run.sh: fix for new command line support

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic commented:
@macpijan makes sense. For PC Engines my goal is different, I would choose more lightweight OS. I want to test this on QEMU. Then, I can move to hardware. Unfortunately, I keep going down the dependency hell, and there is no bottom to it so far.
